### PR TITLE
Remove prover state

### PR DESCRIPTION
**Pull request by mrmr1993**
This PR removes the concept of 'global prover state'. This was effectively never used, and added complexity to the types and the runner code.

### CODE REVIEW

**Review by mimoo — APPROVED** [review]

File: src/base/snark0.ml [hunk 1]
```diff
@@ -1718,31 +1699,19 @@ module Run = struct
...
-                         so we peek one layer up the callstack and 'borrow' the
-                         state from there.
-                         For our purposes, this will always work when the proof
-                         is created using this functor's methods rather than
-                         the checked versions, and so is 'safe enough'.
-                      *)
                       let old_state = !state in
-                      state :=
-                        { state' with prover_state = old_state.prover_state } ;
+                      state := state' ;
```
Comment (jspada) [inline src/base/snark0.ml:1711]: Nice to comment this one (why state flips).

**Review by jspada — APPROVED** [review]
Approved with request for some more comments, if possible.  I checked everything as carefully as possible.

Would be helpful to rename `has_witness` to `generate_the_witness`
and document (if not already done) how interface works with respect to the modes of operation when `generate_the_witness` is set or when there's a constraint system or not.

File: src/base/snark_intf.ml [hunk 2]
```diff
@@ -1103,26 +1064,20 @@ let multiply3 (x : Field.Var.t) (y : Field.Var.t) (z : Field.Var.t)
...
       If a constraint is checked and isn't satisfied, this label will be shown
       in the error message.
   *)
-  val with_label : string -> ('a, 's) Checked.t -> ('a, 's) Checked.t
+  val with_label : string -> 'a Checked.t -> 'a Checked.t
 
   (** Generate the R1CS for the checked computation. *)
   val constraint_system :
-       exposing:((unit, 's) Checked.t, _, 'k_var, _) Data_spec.t
+       exposing:(unit Checked.t, _, 'k_var, _) Data_spec.t
```
[the remark this answers is not in the document]
Comment (jspada) [reply src/base/snark_intf.ml:1077]: Probably out of scope

File: src/base/merkle_tree.ml [hunk 3]
```diff
@@ -151,27 +151,6 @@ let insert hash merge t0 mask0 address x =
 
 let ith_bit n i = (n lsr i) land 1 = 1
```
[the remark this answers is not in the document]
Comment (mrmr1993) [reply src/base/merkle_tree.ml:153]: `update` relied on state updates, which are no longer available. Also, yea, unused!

File: src/base/snark0.ml [hunk 4]
```diff
@@ -2076,23 +2045,15 @@ module Run = struct
       type 'a as_prover = 'a t
 
       let eval_as_prover f =
-        if !(!state.as_prover) && Option.is_some !state.prover_state then (
-          let s = Option.value_exn !state.prover_state in
-          let s, a = f (Runner.get_value !state) s in
-          state := Run_state.set_prover_state (Some s) !state ;
-          a )
+        if !(!state.as_prover) && !state.has_witness then
```
[the remark this answers is not in the document]
Comment (mrmr1993) [reply src/base/snark0.ml:2048]: In OCaml `!` is dereferencing a `ref` (you can think model it as a mutable pointer).

File: snarky_curve/snarky_curve.ml [hunk 5]
```diff
@@ -49,7 +49,7 @@ module type Constant_intf = sig
 end
 
 module type Inputs_intf = sig
-  module Impl : Snarky_backendless.Snark_intf.Run with type prover_state = unit
+  module Impl : Snarky_backendless.Snark_intf.Run
```
[the remark this answers is not in the document]
Comment (mrmr1993) [reply snarky_curve/snarky_curve.ml:52]: I've added a comment, but really this pattern has caused us a lot of issues throughout the code. It only really needs `Field` and `Boolean`, but I don't really want to break out those interfaces from `Snark_intf.Run` in this PR.

**Pull request merged**